In [ ]:
import pandas as pd
import pickle

import pandas as pd
# from sklearn.metrics import mean_squared_log_error
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split

In [ ]:
df = pd.read_csv('https://raw.githubusercontent.com/9394113857/Data-Sets/raghu/heart-data.csv') # FileNotFoundError: [Errno 2] No such file or directory: ''

In [ ]:
df.head() # first 5 rows

age  sex  chest  resting_blood_pressure  serum_cholestoral  \
0   70    1      4                     130                322   
1   67    0      3                     115                564   
2   57    1      2                     124                261   
3   64    1      4                     128                263   
4   74    0      2                     120                269   

   fasting_blood_sugar  resting_electrocardiographic_results  \
0                    0                                     2   
1                    0                                     2   
2                    0                                     0   
3                    0                                     0   
4                    0                                     2   

   maximum_heart_rate_achieved  exercise_induced_angina  oldpeak  slope  \
0                          109                        0      2.4      2   
1                          160                        0      1.6      2   
2                          141                        0      0.3      1   
3                          105                        1      0.2      2   
4                          121                        1      0.2      1   

   number_of_major_vessels  thal    class  
0                        3     3  present  
1                        0     7   absent  
2                        0     7  present  
3                        1     7   absent  
4                        1     3   absent

In [ ]:
# Display the columns of the DataFrame by Sno:
# In this code, we use a for loop to iterate over the columns of the DataFrame (df.columns) along with their corresponding serial numbers (sno).
# We use enumerate with the start=1 argument to begin the serial numbers from 1.
# The print statement formats the output to display the columns with their respective serial numbers in the console.
print("Columns of the DataFrame:")
for sno, column_name in enumerate(df.columns, start=1):
    print(f"{sno:2d} | {column_name}")

Columns of the DataFrame:
 1 | age
 2 | sex
 3 | chest
 4 | resting_blood_pressure
 5 | serum_cholestoral
 6 | fasting_blood_sugar
 7 | resting_electrocardiographic_results
 8 | maximum_heart_rate_achieved
 9 | exercise_induced_angina
10 | oldpeak
11 | slope
12 | number_of_major_vessels
13 | thal
14 | class


In [ ]:
df.rename(columns={"class":"target"},inplace=True)

In [ ]:
# Display the columns of the DataFrame by Sno:
# In this code, we use a for loop to iterate over the columns of the DataFrame (df.columns) along with their corresponding serial numbers (sno).
# We use enumerate with the start=1 argument to begin the serial numbers from 1.
# The print statement formats the output to display the columns with their respective serial numbers in the console.
print("Columns of the DataFrame:")
for sno, column_name in enumerate(df.columns, start=1):
    print(f"{sno:2d} | {column_name}")

Columns of the DataFrame:
 1 | age
 2 | sex
 3 | chest
 4 | resting_blood_pressure
 5 | serum_cholestoral
 6 | fasting_blood_sugar
 7 | resting_electrocardiographic_results
 8 | maximum_heart_rate_achieved
 9 | exercise_induced_angina
10 | oldpeak
11 | slope
12 | number_of_major_vessels
13 | thal
14 | target


In [ ]:
from IPython.display import Markdown, display

# Function to print in bold
def print_bold(text):
    display(Markdown(f"**{text}**"))

# Print the counts of 'absent' and 'present' before replacing
print_bold("Before replacing values of absent and present with 0 and 1:")
print(df['target'].value_counts())


**Before replacing values of absent and present with 0 and 1:**

absent     150
present    120
Name: target, dtype: int64


In [ ]:
df['target'].replace(['absent','present'],[0,1],inplace=True)

In [ ]:
from IPython.display import Markdown, display

# Function to print in bold
def print_bold(text):
    display(Markdown(f"**{text}**"))

# Print the counts of 'absent' and 'present' after replacing with 0 and 1 values:-
print_bold("After replacing:")
print(df['target'].value_counts())


**After replacing:**

0    150
1    120
Name: target, dtype: int64


In [ ]:
df=pd.get_dummies(df)

In [ ]:
df.head()

age  sex  chest  resting_blood_pressure  serum_cholestoral  \
0   70    1      4                     130                322   
1   67    0      3                     115                564   
2   57    1      2                     124                261   
3   64    1      4                     128                263   
4   74    0      2                     120                269   

   fasting_blood_sugar  resting_electrocardiographic_results  \
0                    0                                     2   
1                    0                                     2   
2                    0                                     0   
3                    0                                     0   
4                    0                                     2   

   maximum_heart_rate_achieved  exercise_induced_angina  oldpeak  slope  \
0                          109                        0      2.4      2   
1                          160                        0      1.6      2   
2                          141                        0      0.3      1   
3                          105                        1      0.2      2   
4                          121                        1      0.2      1   

   number_of_major_vessels  thal  target  
0                        3     3       1  
1                        0     7       0  
2                        0     7       1  
3                        1     7       0  
4                        1     3       0

In [ ]:
# doubt
import pandas as pd

# Assuming 'df' is your original DataFrame containing categorical variables
# Apply one-hot encoding using get_dummies()
df = pd.get_dummies(df)

# Save the DataFrame with dummies to a CSV file
df.to_csv('dummies_data.csv', index=False)


In [ ]:
# Read the CSV file back into a DataFrame
df_from_csv = pd.read_csv('dummies_data.csv')

# Calculate the counts of 'absent' and 'present'
absent_count = (df_from_csv['target'] == 0).sum()
present_count = (df_from_csv['target'] == 1).sum()

# Print the counts in the Colab console
print("Count of 'absent':", absent_count)
print("Count of 'present':", present_count)


Count of 'absent': 150
Count of 'present': 120


In [ ]:
x = df.drop('target', axis=1)
y = df['target']


In summary, after executing these two lines of code, x will be a DataFrame containing all the features of the dataset except the 'target' column, and y will be a Series containing the values of the 'target' column.

In [ ]:
# Printing columns for x (DataFrame)
print("Columns of x:")
print(x.columns.tolist())

# Printing column for y (Series)
print("\nColumn of y:")
print(y.name)

Columns of x:
['age', 'sex', 'chest', 'resting_blood_pressure', 'serum_cholestoral', 'fasting_blood_sugar', 'resting_electrocardiographic_results', 'maximum_heart_rate_achieved', 'exercise_induced_angina', 'oldpeak', 'slope', 'number_of_major_vessels', 'thal']

Column of y:
target


1.In the code above, x.columns.tolist() will give you a list of column names from the DataFrame x, and y.name will give you the name of the Series y (which is the name of the column 'target').

2.When you run this code, it will print the columns of x and the column name of y.

In [ ]:
# Printing columns of DataFrame x with a sno column
print("\033[1mSno\tColumn of x:\033[0m")
for idx, col in enumerate(x.columns, start=1):
    print(f"{idx}\t{col}")

# Printing column name of Series y
print("\n\033[1mColumn of y:\033[0m")
print(y.name)


Sno	Column of x:
1	age
2	sex
3	chest
4	resting_blood_pressure
5	serum_cholestoral
6	fasting_blood_sugar
7	resting_electrocardiographic_results
8	maximum_heart_rate_achieved
9	exercise_induced_angina
10	oldpeak
11	slope
12	number_of_major_vessels
13	thal

Column of y:
target


**Points:**

**1.	print("\033[1mSno\tColumn of x:\033[0m"):** This line prints the title "Column of x:" in bold (\033[1m). It also includes a tab (\t) after the title to align the sno column with the DataFrame column names. The "\033[0m" is used to reset the formatting back to normal after printing the title.

**2.	for idx, col in enumerate(x.columns, start=1)::** This line starts a loop to iterate over the column names of DataFrame x. The enumerate() function is used to get both the index (idx) and column names (col) from the DataFrame x.columns. The start=1 parameter sets the starting index value to 1, so the sno column starts with 1.

**3.	print(f"{idx}\t{col}"):** Inside the loop, this line prints the sno (idx) and the column name (col) separated by a tab (\t). This effectively displays the index number followed by the corresponding DataFrame column name.

**4.	print("\n\033[1mColumn of y:\033[0m"):** This line prints the title "Column of y:" in bold. The "\033[1m" is used to start the bold formatting, and "\033[0m" is used to reset the formatting back to normal after printing the title.

**5.	print(y.name):** This line simply prints the name of Series y, which was requested, and since it is the last statement, it will be displayed after the sno column and DataFrame columns.

**Overview:**

Overall, the code prints the columns of DataFrame x in a tabular format with a sno column at the beginning, and the name of Series y is printed separately below the DataFrame columns, all displayed in bold. The sno column helps to identify the index of each column, making it easier to understand the sequence.


In [ ]:
train_x,valid_x,train_y,valid_y=train_test_split(x,y,test_size=0.3,random_state=35)


In [ ]:
logr = LogisticRegression()
logr.fit(train_x,train_y)


/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [ ]:
#new_data=np.array(new_data,dtype='int64')
#new_data=new_data.reshape(1,13)
#xnew_data=pd.DataFrame(new_data)


In [ ]:
pickle.dump(logr,open('heart-model.pkl','wb'))


In [ ]:
model = pickle.load(open('heart-model.pkl','rb'))

In [ ]:
valid_x.head()

age  sex  chest  resting_blood_pressure  serum_cholestoral  \
77    39    0      3                      94                199   
17    53    1      4                     140                203   
133   64    1      4                     120                246   
23    47    1      4                     112                204   
150   41    1      3                     112                250   

     fasting_blood_sugar  resting_electrocardiographic_results  \
77                     0                                     0   
17                     1                                     2   
133                    0                                     2   
23                     0                                     0   
150                    0                                     0   

     maximum_heart_rate_achieved  exercise_induced_angina  oldpeak  slope  \
77                           179                        0      0.0      1   
17                           155                        1      3.1      3   
133                           96                        1      2.2      3   
23                           143                        0      0.1      1   
150                          179                        0      0.0      1   

     number_of_major_vessels  thal  
77                         0     3  
17                         0     7  
133                        1     3  
23                         0     3  
150                        0     3

In [ ]:
result = model.predict(valid_x)
result

array([0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0,
       0, 0, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0,
       1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 0, 0,
       1, 0, 1, 1, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0])

In [ ]:
valid_x.head(1)

age  sex  chest  resting_blood_pressure  serum_cholestoral  \
77   39    0      3                      94                199   

    fasting_blood_sugar  resting_electrocardiographic_results  \
77                    0                                     0   

    maximum_heart_rate_achieved  exercise_induced_angina  oldpeak  slope  \
77                          179                        0      0.0      1   

    number_of_major_vessels  thal  
77                        0     3

In [ ]:
newdata = valid_x.head(1)

In [ ]:
print(newdata)

result2=model.predict(newdata)

print(result)
print(result2)

    age  sex  chest  resting_blood_pressure  serum_cholestoral  \
77   39    0      3                      94                199   

    fasting_blood_sugar  resting_electrocardiographic_results  \
77                    0                                     0   

    maximum_heart_rate_achieved  exercise_induced_angina  oldpeak  slope  \
77                          179                        0      0.0      1   

    number_of_major_vessels  thal  
77                        0     3  
[0 1 1 0 0 1 1 0 0 1 0 1 0 0 0 1 1 1 1 0 0 0 0 0 0 0 1 1 1 1 0 1 0 0 0 1 1
 0 0 1 0 0 0 0 1 1 0 1 1 0 0 0 1 1 0 1 0 1 0 1 1 1 0 1 0 0 1 0 1 1 1 0 1 1
 0 0 0 0 0 0 0]
[0]


In [ ]:
newdata

age  sex  chest  resting_blood_pressure  serum_cholestoral  \
77   39    0      3                      94                199   

    fasting_blood_sugar  resting_electrocardiographic_results  \
77                    0                                     0   

    maximum_heart_rate_achieved  exercise_induced_angina  oldpeak  slope  \
77                          179                        0      0.0      1   

    number_of_major_vessels  thal  
77                        0     3

In [ ]:
result2

array([0])

In [ ]:
# Load the dumped model
heart_model = pickle.load(open('heart-model.pkl', 'rb'))

1.	Enter age: 45
2.	Enter sex (0 for female, 1 for male): 1
3.	Enter cp (chest pain type): 2
4.	Enter trestbps (resting blood pressure): 130
5.	Enter chol (serum cholesterol level): 220
6.	Enter fbs (fasting blood sugar > 120 mg/dl): 0
7.	Enter restecg (resting electrocardiographic results): 1
8.	Enter thalach (maximum heart rate achieved): 160
9.	Enter exang (exercise-induced angina): 0
10.	Enter oldpeak (ST depression induced by exercise relative to rest): 1.5
11.	Enter slope (the slope of the peak exercise ST segment): 2
12.	Enter ca (number of major vessels colored by fluoroscopy): 0
13.	Enter thal (thalassemia): 2

14.	Prediction: [1]



1.	Enter age: 35
2.	Enter sex (0 for female, 1 for male): 0
3.	Enter cp (chest pain type): 0
4.	Enter trestbps (resting blood pressure): 120
5.	Enter chol (serum cholesterol level): 200
6.	Enter fbs (fasting blood sugar > 120 mg/dl): 0
7.	Enter restecg (resting electrocardiographic results): 0
8.	Enter thalach (maximum heart rate achieved): 180
9.	Enter exang (exercise-induced angina): 0
10.	Enter oldpeak (ST depression induced by exercise relative to rest): 0.5
11.	Enter slope (the slope of the peak exercise ST segment): 1
12.	Enter ca (number of major vessels colored by fluoroscopy): 0
13.	Enter thal (thalassemia): 1
14.	Prediction: [0]



In [ ]:
# Take input from the user
age = int(input("Enter age: "))
sex = int(input("Enter sex (0 for female, 1 for male): "))
cp = int(input("Enter chest pain type (0-3): "))
trestbps = int(input("Enter resting blood pressure: "))
chol = int(input("Enter serum cholesterol level: "))
fbs = int(input("Enter fasting blood sugar (0 or 1): "))
restecg = int(input("Enter resting electrocardiographic results (0-2): "))
thalach = int(input("Enter maximum heart rate achieved: "))
exang = int(input("Enter exercise-induced angina (0 or 1): "))
oldpeak = float(input("Enter ST depression induced by exercise relative to rest: "))
slope = int(input("Enter the slope of the peak exercise ST segment (0-2): "))
ca = int(input("Enter number of major vessels (0-3) colored by fluoroscopy: "))
thal = int(input("Enter thalassemia (1-3): "))

# Create a new DataFrame with the user input
user_input = pd.DataFrame(
    {
        'age': [age],
        'sex': [sex],
        'cp': [cp],
        'trestbps': [trestbps],
        'chol': [chol],
        'fbs': [fbs],
        'restecg': [restecg],
        'thalach': [thalach],
        'exang': [exang],
        'oldpeak': [oldpeak],
        'slope': [slope],
        'ca': [ca],
        'thal': [thal]
    }
)

# Make predictions using the loaded model
prediction = heart_model.predict(user_input)

# Print the predicted result
if prediction[0] == 0:
    print("The user is not likely to have heart disease.")
else:
    print("The user is likely to have heart disease.")


Enter age: 45
Enter sex (0 for female, 1 for male): 1
Enter chest pain type (0-3): 2
Enter resting blood pressure: 130
Enter serum cholesterol level: 220
Enter fasting blood sugar (0 or 1): 0
Enter resting electrocardiographic results (0-2): 1
Enter maximum heart rate achieved: 160
Enter exercise-induced angina (0 or 1): 0
Enter ST depression induced by exercise relative to rest: 1.5
Enter the slope of the peak exercise ST segment (0-2): 2
Enter number of major vessels (0-3) colored by fluoroscopy: 0
Enter thalassemia (1-3): 2


ValueError: ignored